src/scenic/core/utils.py

```diff
@@ -216,6 +217,95 @@ def unifyMesh(mesh, verbose=False):
         return mesh
 
 
+def repairMesh(mesh, pitch=(1 / 2) ** 6, verbose=True):
+    """Attempt to repair a mesh, returning a proper 2-manifold.
+
+    Repair is attempted via several steps, each sacrificing more accuracy
+    but with a higher chance of returning a proper volumetric mesh.
+
+    Repair is first attempted with easy fixes, like merging vertices and fixing
+    winding. These will usually not deteriorate the quality of the mesh.
+
+    Repair is then attempted by voxelizing the mesh, filling it, and then running
+    marching cubes on the mesh. This approach is somewhat accurate but always
+    produces solid objects. (This is to be expected since non watertight hollow
+    objects aren't well defined).
+
+    Repair is finally attempted by using the convex hull, which is unlikely to
+    be accurate but is guaranteed to result in a volume.
+    """
+    # If mesh is already a volume, we're done.
+    if mesh.is_volume:
+        return mesh
+
+    ## Trimesh Processing ##
+    processed_mesh = mesh.process(validate=True, merge_tex=True, merge_norm=True).copy()
+
+    if processed_mesh.is_volume:
+        if verbose:
+            warnings.warn("Mesh was repaired via Trimesh process function.")
+
+        return processed_mesh
+
+    if verbose:
+        warnings.warn(
+            "Mesh could not be repaired via Trimesh process function."
+            " attempting voxelization + marching cubes."
+        )
+
+    ## Voxelization + Marching Cubes ##
+    # Extract largest dimension and scale so that it is unit length
+    dims = numpy.abs(processed_mesh.extents)
+    position = processed_mesh.bounding_box.center_mass
+    scale = numpy.max(dims)
+
+    processed_mesh.vertices -= position
+
+    scale_matrix = numpy.eye(4)
+    scale_matrix[:3, :3] /= scale
+    processed_mesh.apply_transform(scale_matrix)
+
+    # Compute new mesh with largest possible pitch
+    curr_pitch = pitch
+
+    while curr_pitch < 1:
+        new_mesh = processed_mesh.voxelized(pitch=curr_pitch).fill().marching_cubes
+
+        if new_mesh.is_volume:
+            if verbose:
+                warnings.warn(
+                    f"Mesh was repaired via voxelization + marching cubes"
+                    f" with pitch {curr_pitch}. Check the output to ensure it"
+                    f" looks reasonable."
+                )
+
+            # Center new mesh
+            new_mesh.vertices -= new_mesh.bounding_box.center_mass
+
+            # Rescale mesh to original size
+            orig_scale = new_mesh.extents / dims
+            scale_matrix = numpy.eye(4)
+            scale_matrix[:3, :3] /= orig_scale
+            new_mesh.apply_transform(scale_matrix)
```

Comment:
This could result in non-uniform scaling, right? Wouldn't it be better to apply the inverse of the `scale_matrix` you used above? (Likewise for the translation.)

Reply:
Hmm, I was thinking that it would be more important that we ensure the bounds are the same as the input mesh, even if it's slightly distorted. I haven't thought about the consequences of this in too much detail though.

Reply:
Per our discussion, let's keep your approach so that the bounds are preserved, even if it means the mesh is slightly distorted. I was concerned the bounds could change dramatically if the voxelization drops thin parts of the mesh, but the default algorithm seems not to do that. In any case you do warn the user to check the result.